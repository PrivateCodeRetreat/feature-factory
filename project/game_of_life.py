class GameOfLife:
    def __init__(self):
        self.alive_cells = set()
        self.cell_ages = {}

    def __str__(self) -> str:
        height = self.get_max_y() + 1
        width = self.get_max_x() + 1
        return "\n".join(["".join([f"{self.get_age(x,y)}" if self.is_alive(x, y) else " " for x in range(width)]) for y in range(height)])
    
    def get_max_x(self):
        return max([cell[0] for cell in self.alive_cells], default=0)
    
    def get_max_y(self):
        return max([cell[1] for cell in self.alive_cells], default=0)

    def set_alive(self, x, y):
        self.alive_cells.add((x, y))
        self.cell_ages[(x, y)] = 1

    def is_alive(self, x, y):
        return (x, y) in self.alive_cells

    def get_age(self, x, y):
        return self.cell_ages.get((x, y), 0)

    def advance(self):
        new_alive_cells = set()
        new_cell_ages = {}
        neighbor_offsets = [(-1, -1), (-1, 0), (-1, 1),
                            ( 0, -1),          ( 0, 1),
                            ( 1, -1), ( 1, 0), ( 1, 1)]

        potential_cells = {neighbor for cell in self.alive_cells for neighbor in 
                           [(cell[0] + dx, cell[1] + dy) for dx, dy in neighbor_offsets]}

        for cell in potential_cells:
            alive_neighbors = sum((neighbor in self.alive_cells) for neighbor in 
                                  [(cell[0] + dx, cell[1] + dy) for dx, dy in neighbor_offsets])
            current_age = self.cell_ages.get(cell, 0)

            if alive_neighbors == 3 or (alive_neighbors == 2 and cell in self.alive_cells):
                new_alive_cells.add(cell)
                new_cell_ages[cell] = current_age + 1

        self.alive_cells = new_alive_cells
        self.cell_ages = new_cell_ages