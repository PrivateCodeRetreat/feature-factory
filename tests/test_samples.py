from approvaltests.approvals import verify, verify_all

from approvaltests.reporters.python_native_reporter import PythonNativeReporter
from approvaltests import Options, verify_as_json
from approvaltests.reporters.generic_diff_reporter_factory import GenericDiffReporterFactory
from approvaltests import set_default_reporter
from approvaltests.storyboard import Storyboard

from game_of_life import GameOfLife

import tkinter as tk

class TkinterGui:

    def __init__(self, game_of_life,width=10,height=10):
        self.width = width
        self.height = height
        self.game = game_of_life
        self.root = tk.Tk()
        self.root.title("Game of Life")
        self.cell_size = 10
        self.frame_duration = 400
        self.cells = {}
        self.create_widgets()
    
    def show(self):
        self.root.after(self.frame_duration, self.advance_turn)
        self.root.mainloop()
    
    def advance_turn(self):
        self.game.advance()
        self.update_cells()
        self.root.after(self.frame_duration, self.advance_turn)

    def create_widgets(self):
        for x in range(self.width):
            for y in range(self.height):
                cell_frame = tk.Frame(self.root, width=self.cell_size, height=self.cell_size,
                                      borderwidth=1, relief='raised')
                cell_frame.grid(row=y, column=x)
                cell = tk.Label(cell_frame, bg='white', width=2, height=1)
                cell.pack(padx=1, pady=1)
                self.cells[(x, y)] = cell
        self.update_cells()
    
    def update_cells(self):
        for x in range(self.width):
            for y in range(self.height):
                if self.game.is_alive(x, y):
                    self.cells[(x, y)].config(bg='blue')
                    self.cells[(x, y)].config(text=self.game.get_age(x,y))
                else:
                    self.cells[(x, y)].config(bg='grey')
                    self.cells[(x, y)].config(text='')

def test_tkinter_gui():
    game = GameOfLife()
    create_glider_at(game, 0, 0)
    gui = TkinterGui(game)
    gui.show()

def create_square_at(game,x,y):
    game.set_alive(x+0,y+0)
    game.set_alive(x+0,y+1)
    game.set_alive(x+1,y+0)
    game.set_alive(x+1,y+1)

def create_glider_at(game,x,y):
    game.set_alive(x+0,y+1)
    game.set_alive(x+1,y+2)
    game.set_alive(x+2,y+0)
    game.set_alive(x+2,y+1)
    game.set_alive(x+2,y+2)

def verify_story_board(game,steps):
    storyboard = Storyboard()
    storyboard.add_frame(game)
    for i in range(steps):
        game.advance()
        storyboard.add_frame(game)
    verify(storyboard)

def test_game_of_life():
    game = GameOfLife()
    create_glider_at(game, 0, 0)
    verify_story_board(game,4)

def test_game_of_life_with_blinker():
    game = GameOfLife()
    game.set_alive(1, 1)
    game.set_alive(1, 2)
    game.set_alive(1, 3)
    game.advance()
    assert game.is_alive(0, 2)
    assert game.is_alive(1, 2)
    assert game.is_alive(2, 2)

def test_square_age():
    game = GameOfLife()
    create_square_at(game,0,0)
    assert game.get_age(3,3) == 0
    assert game.get_age(0,0) == 1
    game.advance()
    assert game.get_age(0,0) == 2