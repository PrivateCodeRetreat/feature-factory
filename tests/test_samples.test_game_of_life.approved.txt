Initial:
  1
1 1
 11

Frame #1:
 1  
  21
 22 

Frame #2:
  1 
   2
 331

Frame #3:
    
 1 3
  42
  1 

Frame #4:
    
   4
 1 3
  21

Frame #5:
     
  1  
   41
  32 

